include/deal.II/numerics/vector_tools_project.templates.h

```diff
@@ -900,44 +940,15 @@ namespace VectorTools
           const Quadrature<dim - 1> &q_boundary,
           const bool                 project_to_boundary_first)
   {
-    if (dim == spacedim)
-      {
-        const Mapping<dim> *const mapping_ptr =
-          dynamic_cast<const Mapping<dim> *>(&mapping);
-        const DoFHandler<dim> *const dof_ptr =
-          dynamic_cast<const DoFHandler<dim> *>(&dof);
-        const Function<dim,
-                       typename VectorType::value_type> *const function_ptr =
-          dynamic_cast<const Function<dim, typename VectorType::value_type> *>(
-            &function);
-        Assert(mapping_ptr != nullptr, ExcInternalError());
-        Assert(dof_ptr != nullptr, ExcInternalError());
```

Comment:
Are the pointers checked internally?

Reply:
These casts were only done to be able to specialize the code from a `dim, spacedim` pair to a code with `dim == spacedim` to be compatible with `MatrixFRee`, and guarded by an `if` statement. So the casts were a workaround, and the new solution with `std::enable_if_t`, covers this in a more systematic way.